Implement maven pom javac release option hint (JEP 247).

mbien opened the pull request:
- converts matching source/target options to the release option
 - supports properties and the standard maven-compiler-plugin
 - offers only hints when source/release is set to at least 9+ (could be improved in future once I figure out how to query the JDK version of the project)

It is fairly basic and has a small cosmetic issue. I don't know how to replace elements in-place, so it simply adds the release element to the end of the list if the hint is applied, no matter where source/target was.

I also don't know how to apply hints in the junit tests. When I tried to do that it complained that the model was read only (org.netbeans.modules.xml.xam.dom.ReadOnlyAccess). I suppose its a dummy model.

How to test:
create a new java project and it should already show the hint. It only looks for properties and the config of the standard javac plugin.

ebarboni commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,217 @@
+@NbBundle.Messages({
+    "TIT_UseReleaseVersionHint=Convert matching source/target javac options to the release option.",
+    "DESC_UseReleaseVersionHint=Matching source/target options can be replaced with a single release option (JEP 247). This enforces API compatibility with the chosen Java release.",
+    "FIX_UseReleaseVersionHint=Convert to release option for strickt compatibility checks."})
```
strickt sounds german

mbien replied:
fixed. Thanks for catching this.

ebarboni commented:
if maven-compiler-plugin version is <3.6 it would not be able to use the release parameter.

mbien commented:
> if maven-compiler-plugin version is <3.6 it would not be able to use the release parameter.

@ebarboni good info! How do I compare versions? Is there a utility for this?

matthiasblaesing commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+            return Collections.emptyList();
+        }
+
+        if (source == target && source >= 9) {
```
I think you could go down to 6 - at least with OpenJDK 11 javac reports:

```
  --release <release>
        Compile for a specific VM version. Supported targets: 6, 7, 8, 9, 10, 11
```

mbien replied:
I actually forgot that it goes lower then 9. But it makes sense since now I remember that I read JDK 20 is going to drop support for 7 and below.

 - JDK 12 dropped support for 6
 - JDK 20 dropped support for 7 -> https://inside.java/2022/09/26/quality-heads-up/

The current implementation will only offer the hint if the values are 9 or greater and match. This implies that the JDK is also 9 or later, which means it is always safe to convert to release.

That is why there is no JDK check. I thought to keep it simple otherwise we would have to update this hint periodically each time a JDK drops a version. You think its worth to cover the EOL versions too with this hint?

mbien replied:
I also wouldn't know how to query the configured JDK. The method has a `Project` parameter which has a lookup. I could get the `NbMavenProject` which has a bit more context but still no JDK.

mbien replied:
i checked all registered services in the lookup of `org.netbeans.api.project.Project` but couldn't find anything which would give me access to the configured JDK of the maven project. Going to shelve the JDK check but would be willing to implement it in a followup PR.

matthiasblaesing commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+            String targetText = parent.getChildElementText(POMQName.createQName(prefix+TARGET_TAG, true));
+            if (isPropertyName(targetText)) {
+                release = targetText;
+                targetText = getProperty(targetText, parent.getModel());
```
Is this a good idea? If I read this correctly, this evaluates a property. My assumption would be, that there is a reason the user does not use the default properties from maven compiler plugin. If so you can't assume that the property will always hold the same value.

mbien replied:
it evaluates the property to check what is in it (9+ check a few lines below). But it remembers the property String. So when it adds the release tag it will use the property again, not the int value.

matthiasblaesing commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+        if (build != null) {
+            for (Plugin plugin : build.getPlugins()) {
+                if (plugin.getArtifactId().equals("maven-compiler-plugin")) {
+                    if (!isPluginCompatible(plugin)) {
```
Should there also be a guard against using this when JDK platform is lower than 9?

mbien commented:
@matthiasblaesing do you have any remaining concerns integrating this in the current form or should we move it to NB17?

mbien commented:
fixed potential NPE when artifactID is null

mbien commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    private List<ErrorDescription> createHintsForParent(String prefix, POMComponent parent) {
+
```
ok, will add the fast path

mbien replied:
you are right. It looks better.

mbien commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+
+        try {
+            String sourceText = parent.getChildElementText(POMQName.createQName(prefix+SOURCE_TAG, true));
+            if (isProperty(sourceText)) {
```
the hint will only show up if there the pom has both, source and target defined and both are set to the same value. It should simply give up if one of the two is not valid or missing.

The main motivation behind this hint was because I met many who still didn't know that the release option exists and why it is better. I personally never set one without the other. Haven't thought about converting a single option into release - you think this would be save in all cases?

sdedic replied:
No, as it only triggers om same strings, it is functionally OK as it is, but the code does not express that intention much clearly :) Maybe an explanation comment that a catch-all-errors handler is OK would be helpful.

mbien replied:
ah i see. Adding a comment to the catch block.

matthiasblaesing commented:
> @matthiasblaesing do you have any remaining concerns integrating this in the current form or should we move it to NB17?

No problem - with the reasoning that 8 is dead and thus release will always be there, we are only missing a corner-case. The approach for the properties seems reasonable.

mbien commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+
+            String targetText = parent.getChildElementText(POMQName.createQName(prefix+TARGET_TAG, true));
+            if (isProperty(targetText)) {
+                release = targetText;
```
i think this doesn't matter, the user will want to rename/cleanup the properties anyway after the hint.
The typical case would have both use the same property - when the dev is not aware release exists.

single property case:
```xml
    <properties>
        <java-version>11</java-version>
        <!-- somewhere later in the compiler plugin -->
        <maven.compiler.source>${java-version}</maven.compiler.source>
        <maven.compiler.target>${java-version}</maven.compiler.target>
    </properties>
```
->
```xml
    <properties>
        <java-version>11</java-version>
        <!-- somewhere later in the compiler plugin -->
        <maven.compiler.release>${java-version}</maven.compiler.release>
    </properties>
```

two property case:
```xml
    <properties>
        <source>11</source>
        <target>11</target>
        <!-- somewhere later in the compiler plugin -->
        <maven.compiler.source>${source}</maven.compiler.source>
        <maven.compiler.target>${target}</maven.compiler.target>
    </properties>
```
->
```xml
    <properties>
        <source>11</source>
        <target>11</target>
        <!-- somewhere later in the compiler plugin -->
        <maven.compiler.release>${target}</maven.compiler.release>
    </properties>
```
cleanup required

mbien commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+
+            source = Integer.parseInt(sourceText);
+            target = Integer.parseInt(targetText);
+            if (release == null) {
```
release can be either a property like "${src}" or a java version like "11". 

If its still null here its not a property, so it parses the target text and assigns it. After this point it has to have a value. Its either an int or the property name.

sdedic replied:
Got it. In the case property ref is used, the release ought use the property tef too. Thanks.

mbien commented on java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java:
```diff
@@ -0,0 +1,241 @@
+            for (POMComponent prop : parent.getChildren()) {
+                String name = prop.getPeer().getNodeName();
+                if (name.equals(prefix+SOURCE_TAG) || name.equals(prefix+TARGET_TAG)) {
+                    hints.add(createHintForComponent(prefix, prop, parent.getModel(), release));
```
release replaces source/target in javac. You can't mix them up:
```
javac -source 11 --release 11
error: option --source cannot be used together with --release
```
So I don't check if there is a release already there (the javac config would be already failing). It would ignore it right now and add another release element to the xml.

mbien replied:
the only reason to compute release is to do the `>= 9` check. Otherwise it would simply collapse source and target to release if they are equal strings or properties which point to equal strings

sdedic replied:
Yes, javac will error when release + source/target is present. But please add the check since users are creative, and may make unintentionally an erroneous pom (source, target AND release). A hint+fix that appears, removes source+target and **duplicates** the release will add even more confusion.

mbien replied:
pick your favorite:
1) should it remove the old release tag if it finds one (and add the new one later)
2) it only removes source/target but doesn't add another release tag if there is one (even if it has the wrong value)

mbien replied:
implementing 1 for now since its fairly elegant to implement

mbien commented:
@neilcsmith-net is master already branched? Or can I still merge?

neilcsmith-net commented:
@mbien you have about 10 minutes! :smile:

mbien commented:
awesome merged. Usually I would wait for approval but I think I addressed all review comments and implemented everything. If there is an issue I can fix it during RC.
